### PR TITLE
Add Integration Tests for SAML ECP Profile

### PR DESCRIPTION
**Pull request by WinmaH**
Reopened https://github.com/wso2/product-is/pull/3953
When should this PR be merged

This PR should me merged after merging the following
wso2-extensions/identity-inbound-auth-saml#196

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java [hunk 1]
```diff
@@ -299,6 +304,15 @@ public void createApplication(SAMLConfig config, String appName) throws Exceptio
         inboundAuthenticationConfig.setInboundAuthenticationRequestConfigs(
                 new InboundAuthenticationRequestConfig[]{requestConfig});
 
+        if(config.httpBinding.equals(HttpBinding.HTTP_SOAP)){
```
Comment (IsuraD) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: formatting issues

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java [hunk 2]
```diff
@@ -166,7 +170,8 @@ public boolean getSetUserClaims() {
     protected enum App {
 
         SUPER_TENANT_APP_WITH_SIGNING("travelocity.com", true),
-        TENANT_APP_WITHOUT_SIGNING("travelocity.com-saml-tenantwithoutsigning", false);
+        TENANT_APP_WITHOUT_SIGNING("travelocity.com-saml-tenantwithoutsigning", false),
+        ECP_APP("https://localhost/shibboleth",false);
```
Comment (IsuraD) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: Change the ECP app name.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java [hunk 3]
```diff
@@ -66,6 +67,8 @@
     private static final String NAMEID_FORMAT = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress";
     private static final String LOGIN_URL = "/carbon/admin/login.jsp";
 
+    protected static final String SAML_ECP_SSO_URL = "https://localhost:9853/samlecp";
+    protected static final String SAML_ECP_ACS_URL = "https://localhost/Shibboleth.sso/SAML2/ECP";
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: Any reason for using "Shibboleth' in the ACS url?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+
+public class SAMLECPSSOTestCase extends AbstractSAMLSSOTestCase {
+
+    private SAMLConfig config;
+    private static final Log log = LogFactory.getLog(AbstractSAMLSSOTestCase.class);
+    private static final String APPLICATION_NAME = "SAML-ECP-TestApplication";
+    private File identityXML;
+    private ServerConfigurationManager serverConfigurationManager;
+    private static final String SAML_ECP_ISSUER = "https://localhost/shibboleth";
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: Please add new line for clarity

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        resetISConfiguration();
+    }
+
+    @Test(description = "Add service provider", groups = "wso2.is", priority = 1)
+    public void testAddSP() throws Exception {
+        Boolean isAddSuccess = ssoConfigServiceClient.addServiceProvider(super.createECPServiceProviderDTO(config));
+        Assert.assertTrue(isAddSuccess, "Adding a service provider has failed for " + config);
+        SAMLSSOServiceProviderDTO[] samlssoServiceProviderDTOs = ssoConfigServiceClient
+                .getServiceProviders().getServiceProviders();
+        log.debug(samlssoServiceProviderDTOs[0].getIssuer());
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: Please add a if clause to check if debug is enabled.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    @Test(alwaysRun = true, description = "Testing SAML ECP login", groups = "wso2.is" , dependsOnMethods = { "testAddSP" })
+    public void testSAMLECPLogin() {
+        try {
+            HttpResponse response;
+            String samlECPReq = buildECPSAMLRequest(SAML_ECP_ACS_URL ,SAML_ECP_ISSUER);
+            response = Utils.sendECPPostRequest(SAML_ECP_SSO_URL,USER_AGENT, httpClient, config.getUser().getUsername() ,config.getUser().getPassword(), samlECPReq );
+            String result = extractDataFromResponse(response);
+            log.debug("This is the Result "+result);
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: Please add a if clause to check if debug is enabled.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+            int responseCode = response.getStatusLine().getStatusCode();
+            log.debug(responseCode);
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: Please add a if clause to check if debug is enabled.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            Assert.assertEquals(responseCode, 200, "Successful login response returned code " + responseCode);
+            Assert.assertTrue(result.contains("urn:oasis:names:tc:SAML:2.0:status:Success"),"Successfully Authenticated the user for "+config);
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: The second parameter in assertTrue is printed when the first parameter gives 'false' as the value.
In that case, you should provide a message which indicates the assertTrue condition failed.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            Assert.assertTrue(result.contains("<SOAP-ENV:Envelope xmlns:SOAP-ENV=\"http://schemas.xmlsoap.org/soap/envelope/\">"),"Successfully retrieved the SAML Response bound with SOAP");
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: The second parameter in assertTrue is printed when the first parameter gives 'false' as the value.
In that case, you should provide a message which indicates the assertTrue condition failed.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+            Assert.assertEquals(responseCode, 500, "Successfully returned a SOAP fault with internal server error " + responseCode);
+        } catch (Exception e) {
+            Assert.fail("SAML ECP SOAP fault test failed for " + config, e);
+        }
+    }
+
+    @Test(description = "Remove service provider", groups = "wso2.is", dependsOnMethods = { "testSOAPFault" })
+    public void testRemoveSP()
+            throws Exception {
+        Boolean isAddSuccess = ssoConfigServiceClient.removeServiceProvider(config.getApp().getArtifact());
```
Comment (denuwanthi) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: isAddSuccess does not imply that you are trying to removeservice provider.
Shall we use a variable like, 'isServiceProviderRemoved' or something which implies the SP removal is happening?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java (hunk 1 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: > formatting issues

fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java (hunk 2 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: > Change the ECP app name.

fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java (hunk 3 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: > Any reason for using "Shibboleth' in the ACS url?

Changed the ECP app ACS URL

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java (hunk 5 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > Please add a if clause to check if debug is enabled.

Removed unnecessary debug logs

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java (hunk 7 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > Please add a if clause to check if debug is enabled.

removed unnecessary debug logs.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java (hunk 6 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > Please add a if clause to check if debug is enabled.

added

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java (hunk 4 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > Please add new line for clarity

Added a new line

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java (hunk 8 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > The second parameter in assertTrue is printed when the first parameter gives 'false' as the value.
> In that case, you should provide a message which indicates the assertTrue condition failed.

Updated the failure message

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    public void testSAMLECPLogin() {
+        try {
+            HttpResponse response;
+            String samlECPReq = buildECPSAMLRequest(SAML_ECP_ACS_URL ,SAML_ECP_ISSUER);
+            response = Utils.sendECPPostRequest(SAML_ECP_SSO_URL,USER_AGENT, httpClient, config.getUser().getUsername() ,config.getUser().getPassword(), samlECPReq );
+            String result = extractDataFromResponse(response);
+            log.debug("This is the Result "+result);
+            int responseCode = response.getStatusLine().getStatusCode();
+            log.debug(responseCode);
+            Assert.assertEquals(responseCode, 200, "Successful login response returned code " + responseCode);
```
[the remark this answers is not in the document]
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > The 3rd parameter should indicate an error message if assert equals fails."Successful login response returned code " does not imply that.

Updated the failure message

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java (hunk 9 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > The second parameter in assertTrue is printed when the first parameter gives 'false' as the value.
> In that case, you should provide a message which indicates the assertTrue condition failed.

Updated the failure message

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java (hunk 10 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/SAMLECPSSOTestCase.java]: > isServiceProviderRemoved

updated the variable name with 'isServiceProviderRemoved'

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java [hunk 12]
```diff
@@ -166,7 +170,8 @@ public boolean getSetUserClaims() {
     protected enum App {
 
         SUPER_TENANT_APP_WITH_SIGNING("travelocity.com", true),
-        TENANT_APP_WITHOUT_SIGNING("travelocity.com-saml-tenantwithoutsigning", false);
+        TENANT_APP_WITHOUT_SIGNING("travelocity.com-saml-tenantwithoutsigning", false),
```
Comment (IsuraD) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: formatting issues.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/util/Utils.java [hunk 13]
```diff
@@ -366,6 +374,24 @@ public static HttpResponse sendGetRequest(String url, String userAgent, HttpClie
...
+        byte[] encodedAuth = Base64.encodeBase64(
+                auth.getBytes(StandardCharsets.UTF_8));
+        String authHeader = "Basic " + new String(encodedAuth);
+        request.setHeader(HttpHeaders.AUTHORIZATION, authHeader);
+        request.setHeader(HttpHeaders.CONTENT_TYPE, "text/xml; charset=utf-8");
+        try {
+            request.setEntity(new StringEntity(soapRequest));
+            response = httpClient.execute(request);
+        } catch (Exception e) {
+            log.error("Error while sending the ECP Request");
```
Comment (IsuraD) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/util/Utils.java]: If an exception occurs, the response will be equal to null. So the subsequent calls will be failed with NullPointer Exceptions.

Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/util/Utils.java]: > If an exception occurs, the response will be equal to null. So the subsequent calls will be failed with NullPointer Exceptions.

fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java (hunk 12 above)
Comment (WinmaH) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/saml/AbstractSAMLSSOTestCase.java]: > formatting issues.

fixed

**Pull request merged**